### PR TITLE
Make all Storage APIs available via Swift

### PR DESCRIPTION
**Pull request by paulb777**
- All Storage APIs available via Swift
- Component model moved to Swift
- Objective C API coverage tests added
- Public API typedefs are kept only for Objective C and CocoaPods via an additional Objective C public header
- APIs that previously returned `StorageHandle` now return `String`
- API doc type references are updated from Objective C to Swift

TODO:

- migrate component unit tests to Swift
- implement previously auto-generated async/await APIs
- API review
- Release Notes
- umbrella header

### CODE REVIEW

File: FirebaseStorage/Tests/Unit/FIRStorageTestHelpers.m [hunk 1]
```diff
@@ -13,8 +13,9 @@
 // limitations under the License.
 
 #import "FirebaseStorage/Tests/Unit/FIRStorageTestHelpers.h"
+#import "FirebaseStorage/Sources/Public/FirebaseStorage/FIRStorage.h"
 
-#import "FirebaseStorage/Sources/FIRStorageComponent.h"
+//#import "FirebaseStorage/Sources/FIRStorageComponent.h"
```
Comment (ncooke3) [inline FirebaseStorage/Tests/Unit/FIRStorageTestHelpers.m:18]: Should this commented out import be removed? 
```suggestion
```

File: FirebaseStorage/Tests/Unit/FIRStorageTests.m [hunk 2]
```diff
@@ -16,7 +16,10 @@
 
 #import "FirebaseStorage/Sources/Public/FirebaseStorage/FIRStorageReference.h"
 
-#import "FirebaseStorage/Sources/FIRStorageComponent.h"
+#ifdef TODO
+// Port the component tests to Swift
+
+//#import "FirebaseStorage/Sources/FIRStorageComponent.h"
```
Comment (ncooke3) [inline FirebaseStorage/Tests/Unit/FIRStorageTests.m:22]: Same here, should this commented out import be removed?
 ```suggestion
```

File: FirebaseStorageSwift/Sources/Storage.swift [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+ * GoogleService-Info.plist.
+ *
+ * If you provide a custom instance of FirebaseApp,
+ * the storage location will be specified via the FirebaseOptions#storageBucket property.
+ */
+@objc(FIRStorage) open class Storage: NSObject {
+  // MARK: - Public APIs
+
+  /**
+   * Creates an instance of FirebaseStorage, configured with the custom FirebaseApp @a app.
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/Storage.swift:38]: What does that `@a` symbol doing there? I have seen it throughout the documentation in this PR.

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import FirebaseAuthInterop
+import FirebaseCore
+import FirebaseCoreInternal
+
+@objc(FIRStorageProvider)
+protocol StorageProvider {
+  @objc func Storage(for bucket: String) -> Storage
+  // TODO: See if we can avoid the `type` parameter by either making it a `Storage` argument to
+  // allow subclasses, or avoid it entirely and fix tests. This was done for StorageCombineUnit,
+  // although we may be able to now port to using `@testable` instead of using the mock.
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageComponent.swift:26]: I'm not sure I 100% follow the TODO but could something like this help?:
```swift 
@objc
protocol StorageProtocol {}

@objc class Storage: NSObject, StorageProtocol {
}

@objc
protocol StorageProvider {
  @objc func storage(for bucket: String) -> StorageProtocol
}
```

`StorageProtocol` could allow for easier fakes.

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  /// The app associated with all Storage instances in this container.
+  private let app: FirebaseApp
+
+  /// A map of active instances, grouped by app. Keys are FirebaseApp names and values are arrays
+  /// containing all instances of Storage associated with the given app.
+  private var instances: [String: Storage] = [:]
+
+  /// Lock to manage access to the instances array to avoid race conditions.
+  private var instancesLock: os_unfair_lock = .init()
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageComponent.swift:40]: should the lock synchronize serially? i.e. using a dispatch queue instead

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      }]
+  }
+
+  // MARK: - StorageProvider conformance
+
+  func Storage(for bucket: String) -> Storage {
+    os_unfair_lock_lock(&instancesLock)
+
+    // Unlock before the function returns.
+    defer { os_unfair_lock_unlock(&instancesLock) }
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageComponent.swift:71]: handy use of `defer {...}`!

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+import Foundation
+import FirebaseAppCheckInterop
+import FirebaseAuthInterop
+import FirebaseCore
+import FirebaseCoreInternal
+
+@objc(FIRStorageProvider)
+protocol StorageProvider {
+  @objc func Storage(for bucket: String) -> Storage
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageComponent.swift:23]: Should this have  lower case method name?
```suggestion
  @objc func storage(for bucket: String) -> Storage
```

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                      ]) { container, isCacheable in
+        guard let app = container.app else { return nil }
+        isCacheable.pointee = true
+        return self.init(app: app)
+      }]
+  }
+
+  // MARK: - StorageProvider conformance
+
+  func Storage(for bucket: String) -> Storage {
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageComponent.swift:67]: same here: lower case method name?
```suggestion
  func storage(for bucket: String) -> Storage {
```

File: FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift [hunk 9]
```diff
@@ -15,10 +15,9 @@
 import FirebaseAuth
 import FirebaseCore
 import FirebaseStorage
-@_exported import FirebaseStorageObjC
 import XCTest
 
-#if swift(>=5.5)
+#if swift(>=5.9)
```
Comment (ncooke3) [inline FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift:20]: 5.9... is this intentionally not supposed to run?

File: FirebaseStorageSwift/Sources/StorageReference.swift [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+    return StorageUploadTask(impl.put(uploadData, metadata: metadata?.impl))
+  }
+
+  /**
+   * Asynchronously uploads data to the currently specified StorageReference.
+   * This is not recommended for large files, and one should instead upload a file from disk.
+   * @param uploadData The Data to upload.
+   * @return An instance of StorageUploadTask, which can be used to monitor or manage the upload.
+   */
+  @objc(putData:) @discardableResult open func __putData(_ uploadData: Data) -> StorageUploadTask {
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageReference.swift:113]: will the style script let give each attribute it's own line?
```suggestion
  @objc(putData:)
  @discardableResult
  open func __putData(_ uploadData: Data) -> StorageUploadTask {
```
Personally, I think it'd be more readable but is not a big deal

File: FirebaseStorageSwift/Sources/StorageReference.swift [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+  }
+
+  @objc override public var description: String {
+    return impl.description
+  }
+
+  /**
+   * Deletes the object at the current path.
+   * @param completion A completion block which returns nil on success, or an error on failure.
+   */
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageReference.swift:388]: A rogue doc comment!

File: FirebaseStorageSwift/Sources/StorageReference.swift [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+        completion(StorageMetadata(impl: impl), error)
+      } else {
+        completion(nil, error)
+      }
+    }
+  }
+
+  // MARK: - Delete
+
+  @objc(deleteWithCompletion:)
```
Comment (ncooke3) [inline FirebaseStorageSwift/Sources/StorageReference.swift:359]:
```suggestion
 /**
  * Deletes the object at the current path.
  * @param completion A completion block which returns nil on success, or an error on failure.
  */
  @objc(deleteWithCompletion:)
```

File: FirebaseStorage/Sources/FIRStorageDeleteTask.h [hunk 13]
```diff
@@ -23,9 +23,9 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  * Task which provides the ability to delete an object in Firebase Storage.
  */
-@interface FIRStorageDeleteTask : FIRStorageTask <FIRStorageTaskManagement>
+@interface FIRStorageDeleteTask : FIRIMPLStorageTask <FIRIMPLStorageTaskManagement>
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorage/Sources/FIRStorageDeleteTask.h:26]: Internal class (not public)

File: FirebaseStorage/Sources/FIRStorageErrors.h [hunk 14]
```diff
@@ -31,23 +31,23 @@ NS_ASSUME_NONNULL_BEGIN
 /**
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorage/Sources/FIRStorageErrors.h:31]: Internal class (not public)

File: FirebaseStorage/Sources/FIRStorageGetDownloadURLTask.h [hunk 15]
```diff
@@ -23,9 +23,9 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  * Task which provides the ability to get a download URL for an object in Firebase Storage.
  */
-@interface FIRStorageGetDownloadURLTask : FIRStorageTask <FIRStorageTaskManagement>
+@interface FIRStorageGetDownloadURLTask : FIRIMPLStorageTask <FIRIMPLStorageTaskManagement>
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorage/Sources/FIRStorageGetDownloadURLTask.h:26]: Internal class (not public)

File: FirebaseStorage/Sources/FIRStorageUpdateMetadataTask.h [hunk 16]
```diff
@@ -23,12 +23,12 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  * Task which provides the ability update the metadata on an object in Firebase Storage.
  */
-@interface FIRStorageUpdateMetadataTask : FIRStorageTask <FIRStorageTaskManagement>
+@interface FIRStorageUpdateMetadataTask : FIRIMPLStorageTask <FIRIMPLStorageTaskManagement>
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorage/Sources/FIRStorageUpdateMetadataTask.h:26]: Internal class (not public)

File: FirebaseStorageSwift/Sources/AsyncAwait.swift [hunk 17]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-@_exported import FirebaseStorageObjC
+import Foundation
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorageSwift/Sources/AsyncAwait.swift:15]: [AsyncAwait.swift](https://github.com/firebase/firebase-ios-sdk/pull/9443/files/1b5d6c4ada1d08582ac26d06b0e4824c73d6f131#diff-5e3be542b270ab77890ef7b3d1cff04f167ac69940fdc37927d43d0517c51035) does not need FirebaseStorageObjC!

File: FirebaseStorageSwift/Sources/StorageComponent.swift (hunk 4 above)
Comment (ncooke3) [reply FirebaseStorageSwift/Sources/StorageComponent.swift:26]: cc: @ryanwilson

File: FirebaseStorageSwift/Sources/StorageComponent.swift (hunk 7 above)
Comment (ncooke3) [reply FirebaseStorageSwift/Sources/StorageComponent.swift:23]: cc: @ryanwilson

File: FirebaseStorageSwift/Sources/StorageComponent.swift (hunk 8 above)
Comment (ncooke3) [reply FirebaseStorageSwift/Sources/StorageComponent.swift:67]: cc: @ryanwilson

File: FirebaseStorageSwift/Sources/StorageComponent.swift [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    os_unfair_lock_lock(&instancesLock)
+
+    // Unlock before the function returns.
+    defer { os_unfair_lock_unlock(&instancesLock) }
+
+    if let instance = instances[bucket] {
+      return instance
+    }
+    let newInstance = FirebaseStorage.Storage(app: app, bucket: bucket)
+    instances[bucket] = newInstance
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply FirebaseStorageSwift/Sources/StorageComponent.swift:77]: cc: @ryanwilson

File: FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift (hunk 9 above)
Comment (ncooke3) [reply FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift:20]: intentional so they dont run

File: FirebaseStorage/Sources/FIRStorage.m [hunk 19]
```diff
@@ -39,7 +38,7 @@
     NSMutableDictionary<NSString * /* bucket */, GTMSessionFetcherService *> *> *_fetcherServiceMap;
 static GTMSessionFetcherRetryBlock _retryWhenOffline;
 
-@interface FIRStorage () {
+@interface FIRIMPLStorage () {
```
[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseStorage/Sources/FIRStorage.m:41]: Yes. The files should be renamed. Also the directories should be renamed to match the library like FirebaseStorageInternal.

However, I'd like to wait until these files are back on master to simplify merging.

File: FirebaseStorage/Tests/Unit/FIRStorageTestHelpers.m (hunk 1 above)
Comment (paulb777) [reply FirebaseStorage/Tests/Unit/FIRStorageTestHelpers.m:18]: Yes.

File: FirebaseStorage/Tests/Unit/FIRStorageTests.m (hunk 2 above)
Comment (paulb777) [reply FirebaseStorage/Tests/Unit/FIRStorageTests.m:22]: This whole file needs to be ported to Swift

File: FirebaseStorageSwift/Sources/StorageComponent.swift (hunk 8 above)
Comment (paulb777) [reply FirebaseStorageSwift/Sources/StorageComponent.swift:67]: Yes. Fixed.

**Comment by paulb777** [comment]
I'm going to merge this here into storage-v9 to make the follow up task PRs more manageable. Feel free to continue to review here or in the storage-v9 umbrella PR - #9355 and I'll address.

**Pull request merged**